File: Python/demo_gtm.py
# -*- coding: utf-8 -*- 
#%reset -f
"""
@author: Hiromasa Kaneko
"""

# Demonstration of GTM

# settings
shapeofmap = [10, 10]
shapeofrbfcenters = [5, 5]
varianceofrbfs = 4
lambdainemalgorithm = 0.001
numberofiterations = 200
#numberofiterations = 10
desplayflag = 1

from sklearn.datasets import load_iris
from gtm import gtm
import matplotlib.pyplot as plt
import matplotlib.figure as figure

# load an iris dataset
iris = load_iris()
#inputdataset = pd.DataFrame(iris.data, columns=iris.feature_names)
inputdataset = iris.data
color = iris.target

# autoscaling
inputdataset = (inputdataset - inputdataset.mean(axis=0)) / inputdataset.std(axis=0,ddof=1)

# construct GTM model
model = gtm( shapeofmap, shapeofrbfcenters, varianceofrbfs, lambdainemalgorithm, numberofiterations, desplayflag)
model.fit(inputdataset)

if model.successflag:
    # calculate of responsibilities
    responsibilities = model.responsibility(inputdataset)
    
    # plot the mean of responsibilities
    means = responsibilities.dot( model.mapgrids )
    plt.figure(figsize=figure.figaspect(1))
    plt.scatter( means[:,0], means[:,1], c=color)
    plt.ylim(-1.1,1.1)
    plt.xlim(-1.1,1.1)
    plt.xlabel("z1 (mean)")
    plt.ylabel("z2 (mean)")
    plt.show()
    
    # plot the mode of responsibilities
    modes = model.mapgrids[responsibilities.argmax(axis=1), :]
    plt.figure(figsize=figure.figaspect(1))
    plt.scatter( modes[:,0], modes[:,1], c=color)
    plt.ylim(-1.1,1.1)
    plt.xlim(-1.1,1.1)
    plt.xlabel("z1 (mode)")
    plt.ylabel("z2 (mode)")
    plt.show()

File: MATLAB/demo_gtm.m
% Demonstration of GTM
clear; close all;

% settings
shapeofmap = [ 10, 10 ];
shapeofrbfcenters = [ 5, 5 ];
varianceofrbfs = 4;
lambdainemalgorithm = 0.001;
numberofiterations = 300;
displayflag = 1;

% load an iris dataset
load fisheriris
inputdataset = meas;

% autoscaling
inputdataset = zscore(inputdataset,1);

% construct GTM model
model = calc_gtm(inputdataset, shapeofmap,shapeofrbfcenters, varianceofrbfs, lambdainemalgorithm, numberofiterations, displayflag);

if model.successflag
    % calculate responsibility
    responsibilities = calc_responsibility(model, inputdataset);

    % plot the mean of the responsibility
    means = responsibilities * model.mapgrids;
    figure; gscatter( means(:,1), means(:,2), species);
    axis square; 
    xlabel( 'z_1 (mean)' ,  'FontSize' , 18 , 'FontName','Meiryo UI');
    ylabel( 'z_2 (mean)' ,  'FontSize' , 18 , 'FontName','Meiryo UI');
    set(gcf, 'Color' , 'w' ); 
    set(gca, 'FontSize' , 18);
    set(gca, 'FontName','Meiryo UI');
    axis([-1.1 1.1 -1.1 1.1]);

    % plot the mode of the responsibility
    [~, maxindex] = max(responsibilities, [], 2);
    modes = model.mapgrids(maxindex, :);
    figure; gscatter( modes(:,1), modes(:,2), species);
    axis square;
    xlabel( 'z_1 (mode)' ,  'FontSize' , 18 , 'FontName','Meiryo UI');
    ylabel( 'z_2 (mode)' ,  'FontSize' , 18 , 'FontName','Meiryo UI');
    set(gcf, 'Color' , 'w' ); 
    set(gca, 'FontSize' , 18);
    set(gca, 'FontName','Meiryo UI');
    axis([-1.1 1.1 -1.1 1.1]);
end
